src/MacroExpander.js

```diff
@@ -0,0 +1,60 @@
+/**
+ * This file contains the “gullet” where macros are expanded
+ * until only non-macro tokens remain.
+ */
+
+var Lexer = require("./Lexer");
+
+function MacroExpander(input, macros) {
+    this.lexer = new Lexer(input);
+    this.macros = macros || {};
+    this.stack = []; // contains tokens in REVERSE order
+    this.discardedWhiteSpace = [];
+}
+
+MacroExpander.prototype.nextToken = function() {
```

Comment:
A comment is definitely a good idea. I think the term “atom” is already used with a different meaning inside the parser. I'm not sure what the correct TeX term for a non-expandable piece of input is, perhaps I should look that up.

I guess I'd rather not expand everything at once because one day we will get definitions. And for that to work, we need to change the set of known macros before we continue parsing. Defining macros is technically inside the gullet, I guess, so it would be possible to have some API which returns the whole expanded token stream. But implementing that is easier if we have *some* point where the tokens pass through one at a time, and definitions can take effect for subsequent tokens. And since the parser consumes tokens one at a time, there is little point in having a “convert-everything-at-once” layer in there at all.

Plus it might happen that in order to get scoping right, i.e. the distinction between `\def` and `\gdef`, we would one day welcome help from the parser to tell us when one group ends so that we can close its scope. I don't say that we should take this approach, but having it as an option feels valuable somehow.